src/Oscoin/Storage/Ledger.hs

```diff
@@ -0,0 +1,282 @@
+    case maybeState of
+        Just st -> pure (block, st)
+        Nothing -> throwError (StateNotInCache $ blockHash block)
+
+getState
+    :: Ledger crypto seal tx output state m
+    -> Crypto.Hashed crypto state
+    -> m (Maybe state)
+getState Ledger{stateStore} stateHash = lookupHashContent stateStore stateHash
+
+
+-----------------------------------------------------------
+-- * Mutations
+-----------------------------------------------------------
+
+-- | Builds an unsealed block from a list of transactions applied to
+-- the state of the most recent block.
+--
+-- This uses 'Oscoin.Crypto.Blockchain.Eval' internally which means
+-- that invalid transactions are disregarded.
+--
+-- Used by 'Oscoin.Consensus.Mining.mineBlock'.
+buildNextBlock
+    :: ( Monad m
+       , Serialise tx
+       , Crypto.Hashable crypto tx
+       , Crypto.Hashable crypto state
+       , Crypto.Hashable crypto (BlockHeader crypto Unsealed)
+       , AuthTree.MerkleHash (Crypto.Hash crypto)
+       )
+    => Ledger crypto seal tx output state m
+    -> Timestamp
+    -> [tx]
+    -> m (Either (LedgerError crypto) (Block crypto tx Unsealed))
+buildNextBlock ledger time txs = runExceptT $ do
+    (tip, tipState) <- ExceptT $ getLatestBlockWithState ledger
+    let (newBlock, newState, receipts) = buildBlock (ledgerEval ledger) time tipState txs (blockHash tip)
+    lift $ do
+        storeHashContent (stateStore ledger) newState
+        forM_ receipts $ ReceiptStore.storeReceipt (receiptStore ledger)
+    pure newBlock
+
+
+-- | Computes the state and receipts associated with the given block
+-- and adds them to the storage.
+--
+-- @initialState@ is the state associated with the parent block of @blk@.
+materializeBlock
+    :: (Monad m, Crypto.Hashable crypto tx)
+    => Ledger crypto seal tx output state m
+    -> state
+    -> Block crypto tx (Sealed crypto seal)
+    -> m state
+materializeBlock ledger prevState blk = do
+    let (newState, receipts) = evalBlock (ledgerEval ledger) prevState blk
+    storeHashContent (stateStore ledger) newState
+    forM_ receipts (ReceiptStore.storeReceipt (receiptStore ledger))
+    pure newState
+
+
+-- | Computes all states and receipts from the blocks in the
+-- blockchain and adds them to the storage.
+--
+-- The genesis block is applied to @initialState@
+materialize
+    :: (Monad m, Crypto.Hashable crypto tx)
+    => Ledger crypto seal tx output state m
+    -> state
+    -> m ()
+materialize ledger initialState = do
+    genesisBlock <- BlockStore.getGenesisBlock (blockStoreReader ledger)
+    allFollowingBlocks <- BlockStore.getBlocksByParentHash (blockStoreReader ledger) (blockHash genesisBlock)
+    _ <- foldlM (materializeBlock ledger) initialState (genesisBlock : Chrono.toOldestFirst (Chrono.reverse allFollowingBlocks))
+    pure ()
+
+
+-- | Adds a block to the end of the blockchain and adds the
+-- corresponding state and receipts.
```

Comment:
Is this really true? Is it not possible that the block is added, but not as a new tip?

Reply:
See #490